runtime/sema/check_composite_declaration.go

```
@@ -1902,6 +1931,19 @@ func (checker *Checker) checkSpecialFunction(
 	}
 }
 
+func (checker *Checker) updateMemberPurity(function *ast.FunctionDeclaration, selfType NominalType) {
+	fnType := checker.Elaboration.FunctionDeclarationFunctionTypes[function]
+	member, present := selfType.MemberMap().Get(function.Identifier.Identifier)
+	if present {
+		// members resolvers are created before the purity analysis is performed, so we update
+		// unresolved purities with the correct values
+		fnMember, ok := member.TypeAnnotation.Type.(*FunctionType)
+		if ok {
+			fnMember.Purity = fnType.Purity
+		}
+	}
```

Fail early, e.g. `if !present { return }` (or panic?), `if !ok { return }`, ...